### DISCUSSION TITLE
Valet doesn't work for previous php version after upgrading php. Failed to open stream: Operation not permitted

### DISCUSSION THREAD

**Discussion by kennethjayden** [laravel/valet#1363] [Q&A] [answered]
My existing version was php8.1 using Laravel Valet 3.3.2. I did a brew php install and the php version was updated to php8.2

The site loaded fine on php8.2. When I switched to use the older version using

```
valet use php@8.1
```

I get this error

```
Warning: require(/MY_LARAVEL_PROJECT/public/index.php): Failed to open stream: Operation not permitted in /MY_DIRECTORY_NAME/.composer/vendor/laravel/valet/server.php on line 250

Fatal error: Uncaught Error: Failed opening required '/MY_LARAVEL_PROJECT/public/index.php' (include_path='.:/usr/local/Cellar/php@8.1/8.1.15/share/php@8.1/pear') in /MY_DIRECTORY_NAME/.composer/vendor/laravel/valet/server.php:250 Stack trace: #0 {main} thrown in /MY_DIRECTORY_NAME/.composer/vendor/laravel/valet/server.php on line 250
```

However, these scenarios below worked fine

```
valet use php@8.0
```

```
valet use php
```

```
valet use php@8.1
valet stop
php artisan serve
```

For this last one, the site could work if I used artisan serve instead of valet and loaded the page via 127.0.0.1:8000

I have tried most of the solutions I could find online. The most comprehensive being to uninstall and reinstall everything related to valet.

```
valet uninstall --force
sudo brew services stop nginx
sudo brew services stop php
sudo brew services stop dnsmasq
brew uninstall nginx
brew uninstall composer
brew uninstall php
brew uninstall php@8.1
brew uninstall php@8.0
brew uninstall dnsmasq
rm -rf ~/.config/valet
```

And then reinstall php, composer and valet. I tried this a few times using different combinations and ensuring the config files are removed when prompted by brew. 

I've also tried setting the public folder and index.php file to 777.

The problem still persisted. 

On console, when doing

```
php -v
```

I get

```
PHP 8.1.15 (cli) (built: Feb  3 2023 12:44:27) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.15, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.15, Copyright (c), by Zend Technologies
```

So it seems that this error occurs under the scenario when valet uses php8.1, which was my previous default php version. Probably some config or links related to valet that was not updated / removed.

I would appreciate if anyone can point me in any direction to investigate further. Thanks in advance!

**Comment by AsmitNepali** [comment]
I'm facing the same issue

**Reply by drbyte** [reply]
What version of Valet?
What version of PHP "before", and what PHP version "after"?

**Reply by AsmitNepali** [reply]
Previously I used 8.0 and I updated to 8.2

Anyway, I fixed that after reinstalling and allowing for disk access.

**Comment by drbyte** [comment, marked as answer]
When switching PHP versions, particularly when downgrading, it is wise to rebuild the global composer vendor dependencies.

```sh
composer global update
```

Also useful whenever you want to switch your command-line version temporarily is to insert the specific PHP version at the beginning of your environment PATH, so that all `php` calls use that version, overriding whatever the main default is set to:
```sh
export PATH="$(brew --prefix php@8.2)/bin:$PATH"
```
(replacing `8.2` with whatever your current terminal session needs, for whatever project you're working on)
